libsolidity/ast/Types.cpp

```diff
@@ -2140,30 +2140,25 @@ FunctionType const* ContractType::newExpressionType() const
 	return m_constructorType;
 }
 
-std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables(std::optional<DataLocation> _location) const
+std::vector<std::tuple<VariableDeclaration const*, u256, unsigned>> ContractType::stateVariables(DataLocation _location) const
```

Comment:
What I meant was:

`ContractType::stateVariables()` -> return *all* state variables (transient *and* plain storage), i.e. `_location = std::nullopt` by default.
`ContractType::stateVariables(DataLocation::Storage)` -> storage state variables only
`ContractType::stateVariables(DataLocation::Transient)` -> transient storage  state variables only

Reply:
I will say though - making `_location` mandatory (as you've done) is also a good option, since it forces to double check every `stateVariables()` callsite. The only annoying thing is the cumbersome usage if you want *all* storage variables with the `for (auto location : {DataLocation::Transient, DataLocation::Storage}) {}`.

Perhaps having it accept a vector of data locations, to move the loop into `ContractType::stateVariables`? Or actually, just leave it as and focus on getting the tests fixed, this whole thing is fairly low on the list of priorities :D

Reply:
I was considering adding the overload with no arguments to return all variables. But then, it is just as before when we had the optional parameter, so it wouldn't really make sense changing it. I agree that having to write the loop may be a little bit cumbersome, but at least it makes everything very obvious.
But we can surely change this at later time.